# （三）手写 Promise

## 1. 构造函数实现

使用 `IIFI` 模块化，`function` 构造函数手写 Promise。
```js
/*
  * 自定义 Promise 函数模块：IIFE立即执行函数
  * 关键：then方法实现
  * 构造函数实现
*/
(function (window) {

  const PENDING = 'pending'
  const RESOLVED = 'fulfilled'
  const REJECTED = 'rejected'

  /*
    Promise 构造函数
    executor: 执行器函数，立即同步执行
  */
  function Promise(executor) {
    const that = this
    // Promise对象状态属性，初始值为pending
    that.status = PENDING
    // 存储结果数据
    that.data = 'undefined'
    // 保存待执行的回调函数，数据结构：{onResolved(){},onRejected(){}}
    that.callbacks = []

    function resolve(value) {
      // 若当前状态不是 pending，直接结束
      if (that.status !== PENDING) {
        return
      }
      // 修改状态
      that.status = RESOLVED
      // 修改值
      that.data = value
      // 如有待执行的 callback 函数，立即异步执行回调函数 onResolved
      if (that.callbacks.length > 0) {
        // 模拟异步执行所有成功的回调函数
        setTimeout(() => {
          that.callbacks.forEach(callbacksObj => {
            callbacksObj.onResolved(value)
          });
        })

      }
    }

    function reject(reason) {
      // 若当前状态不是 pending，直接结束
      if (that.status !== PENDING) {
        return
      }
      // 修改状态
      that.status = REJECTED
      // 修改值
      that.data = reason
      // 如有待执行的 callback 函数，立即异步执行回调函数 onRejected
      if (that.callbacks.length > 0) {
        // 模拟异步执行所有成功的回调函数
        setTimeout(() => {
          that.callbacks.forEach(callbacksObj => {
            callbacksObj.onRejected(reason)
          });
        })

      }
    }

    // 立即同步执行
    try {
      executor(resolve, reject)
    } catch (error) {
      // 若执行器异常，promise 对象变为 rejected 状态
      reject(error)
    }

  }

  /*
    Promise原型对象 then 方法，
    两个回调函数 成功 onResolved ，失败onRejected
    返回一个新的Promise对象
    返回promise的结果由onResolved/onRejected执行结果决定
  */
  Promise.prototype.then = function (onResolved, onRejected) {

    const that = this

    onResolved = typeof onResolved === 'function' ? onResolved : value => value
    onRejected = typeof onRejected === 'function' ? onRejected : reason => { throw reason }

    // 返回一个新的 promise
    return new Promise((resolve, reject) => {
      // 使用指定函数处理，根据执行结果，改变return的promise的状态
      function handle(callback) {
        /* 1.回调函数抛出异常，则返回的Promise就会失败，reason就是error
          2. 回调函数返回的不是 promise，则返回的promise就会成功，value就是result
          3.回调函数返回的是promise，则返回的promise取决于这个promise
        */
        try {
          const result = callback(that.data)
          if (result instanceof Promise) {
            // 3.回调函数返回的是promise，则返回的promise取决于这个promise
            result.then(
              value => resolve(value), // 当result成功，返回的promise也成功
              reason => reject(reason) // 当result失败，返回的promise也失败
            )
            // result.then(resolve, reject) // 等同写法
          } else {
            //2. 回调函数返回的不是 promise，则返回的promise就会成功，value就是result
            resolve(result)
          }
        } catch (error) {
          // 1.回调函数抛出异常，则返回的Promise就会失败，reason就是error
          reject(error)
        }
      }

      // pending状态，保存回调函数
      if (that.status === PENDING) {
        that.callbacks.push({
          onResolved(value) {
            handle(onResolved)
          },
          onRejected(reason) {
            handle(onRejected)
          }
        })
        // resolved状态，异步执行回调函数，改变return的promise的状态
      } else if (this.status === RESOLVED) {
        setTimeout(() => {
          handle(onResolved)
        })
      } else { // 'rejected'，异步执行回调函数，改变return的promise的状态
        setTimeout(() => {
          handle(onRejected)
        })
      }
    })

  }

  /*
    Promise原型对象 catch ,参数为失败的回掉函数 onRejected
    返回一个新的Promise对象
  */
  Promise.prototype.catch = function (onRejected) {
    return this.then(undefined, onRejected)
  }


  /*
    Promise函数对象的 resolve 方法
    返回一个新的Promise对象,Promise.resolve()中可以传入Promise
    返回新promise得结果：
     - 若非promise，则返回成功的promise，value就是这个传入的参数
     - 若为promise，则返回成功或失败的promsie，取决于传入的promise的成功或失败
  */
  Promise.resolve = function (value) {

    return new Promise((resolve, reject) => {
      // 若为promise，则返回成功或失败的promsie，取决于传入的promise的成功或失败
      if (value instanceof Promise) {
        value.then(resolve, reject)
      } else { // 若非promise，则返回成功的promise，value就是这个传入的参数
        resolve(value)
      }
    })

  }

  /*
  Promise函数对象的 reject 方法
   返回一个新的Promise对象 Promise.reject中不能再传入Promise
 */
  Promise.reject = function (reason) {
    return new Promise((resolve, reject) => {
      reject(reason)
    })
  }

  /*
    Promise函数对象的 all 方法,接受一个promise类型的数组
    返回一个新的Promise对象
    所有promise成功则返回成功，只要有一个失败就失败
  */
  Promise.all = function (promises) {
    // 成功的promise的值
    const values = new Array(promises.length)
    // 记录成功的promise的数目
    let count = 0
    return new Promise((resolve, reject) => {
      // 遍历promises，获取每个promise的结果
      promises.forEach((p, index) => {
        // 使用  Promise.resolve(p) 再包装一遍，以接收字面量参数
        Promise.resolve(p).then(
          value => {
            count++
            values[index] = value
            // 全部成功，返回所有成功的值的数组
            if (count === promises.length) {
              resolve(values)
            }

          },
          reason => {
            // 出现失败就返回失败
            resolve(reason)
          }
        )
      })

    })
  }

  /*
    Promise函数对象的 race 方法,接受一个promise类型的数组
    返回一个新的Promise对象
  */
  Promise.race = function (promises) {
    return new Promise((resolve, reject) => {
      promises.forEach(p => {
        Promise.resolve(p).then(
          value => {
            resolve(value)
          },
          reason => {
            reject(reason)
          }
        )
      })
    })
  }

  /*
    resolveDelay：扩展工具方法，延迟返回结果
  */
  Promise.resolveDelay = function (value, time) {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        if (value instanceof Promise) {
          value.then(resolve, reject)
        } else {
          resolve(value)
        }
      }, time)
    })
  }

  /*
    rejectDelay：扩展工具方法，延迟返回结果
  */
  Promise.rejectDelay = function (reason, time) {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        reject(reason)
      }, time)
    })
  }

  // 向外暴露 Promise 函数
  window.Promise = Promise
})(window)
```

## 2. ES6 class 类实现

```js
/*
  * 自定义 Promise 函数模块：IIFE立即执行函数
  * 关键：then方法实现 
  * class类实现
*/
(function (window) {
  const PENDING = 'pending'
  const RESOLVED = 'fulfilled'
  const REJECTED = 'rejected'

  /*
    Promise 构造函数
    executor: 执行器函数，立即同步执行
  */
  class Promise {
    constructor(executor) {
      const that = this
      // Promise对象状态属性，初始值为pending
      that.status = PENDING
      // 存储结果数据
      that.data = 'undefined'
      // 保存待执行的回调函数，数据结构：{onResolved(){},onRejected(){}}
      that.callbacks = []

      function resolve(value) {
        // 若当前状态不是 pending，直接结束
        if (that.status !== PENDING) {
          return
        }
        // 修改状态
        that.status = RESOLVED
        // 修改值
        that.data = value
        // 如有待执行的 callback 函数，立即异步执行回调函数 onResolved
        if (that.callbacks.length > 0) {
          // 模拟异步执行所有成功的回调函数
          setTimeout(() => {
            that.callbacks.forEach(callbacksObj => {
              callbacksObj.onResolved(value)
            });
          })

        }
      }

      function reject(reason) {
        // 若当前状态不是 pending，直接结束
        if (that.status !== PENDING) {
          return
        }
        // 修改状态
        that.status = REJECTED
        // 修改值
        that.data = reason
        // 如有待执行的 callback 函数，立即异步执行回调函数 onRejected
        if (that.callbacks.length > 0) {
          // 模拟异步执行所有成功的回调函数
          setTimeout(() => {
            that.callbacks.forEach(callbacksObj => {
              callbacksObj.onRejected(reason)
            });
          })

        }
      }

      // 立即同步执行
      try {
        executor(resolve, reject)
      } catch (error) {
        // 若执行器异常，promise 对象变为 rejected 状态
        reject(error)
      }

    }
    /*
      Promise原型对象 then 方法，
      两个回调函数 成功 onResolved ，失败onRejected
      返回一个新的Promise对象
      返回promise的结果由onResolved/onRejected执行结果决定
    */
    then(onResolved, onRejected) {

      const that = this

      onResolved = typeof onResolved === 'function' ? onResolved : value => value
      onRejected = typeof onRejected === 'function' ? onRejected : reason => { throw reason }

      // 返回一个新的 promise
      return new Promise((resolve, reject) => {
        // 使用指定函数处理，根据执行结果，改变return的promise的状态
        function handle(callback) {
          /* 1.回调函数抛出异常，则返回的Promise就会失败，reason就是error
            2. 回调函数返回的不是 promise，则返回的promise就会成功，value就是result
            3.回调函数返回的是promise，则返回的promise取决于这个promise
          */
          try {
            const result = callback(that.data)
            if (result instanceof Promise) {
              // 3.回调函数返回的是promise，则返回的promise取决于这个promise
              result.then(
                value => resolve(value), // 当result成功，返回的promise也成功
                reason => reject(reason) // 当result失败，返回的promise也失败
              )
              // result.then(resolve, reject) // 等同写法
            } else {
              //2. 回调函数返回的不是 promise，则返回的promise就会成功，value就是result
              resolve(result)
            }
          } catch (error) {
            // 1.回调函数抛出异常，则返回的Promise就会失败，reason就是error
            reject(error)
          }
        }

        // pending状态，保存回调函数
        if (that.status === PENDING) {
          that.callbacks.push({
            onResolved(value) {
              handle(onResolved)
            },
            onRejected(reason) {
              handle(onRejected)
            }
          })
          // resolved状态，异步执行回调函数，改变return的promise的状态
        } else if (this.status === RESOLVED) {
          setTimeout(() => {
            handle(onResolved)
          })
        } else { // 'rejected'，异步执行回调函数，改变return的promise的状态
          setTimeout(() => {
            handle(onRejected)
          })
        }
      })

    }
    /*
      Promise原型对象 catch ,参数为失败的回掉函数 onRejected
      返回一个新的Promise对象
    */
    catch(onRejected) {
      return this.then(undefined, onRejected)
    }

    /*
    Promise函数对象的 resolve 方法
    返回一个新的Promise对象,Promise.resolve()中可以传入Promise
    返回新promise得结果：
    - 若非promise，则返回成功的promise，value就是这个传入的参数
    - 若为promise，则返回成功或失败的promsie，取决于传入的promise的成功或失败
  */
    static resolve = function (value) {

      return new Promise((resolve, reject) => {
        // 若为promise，则返回成功或失败的promsie，取决于传入的promise的成功或失败
        if (value instanceof Promise) {
          value.then(resolve, reject)
        } else { // 若非promise，则返回成功的promise，value就是这个传入的参数
          resolve(value)
        }
      })

    }

    /*
Promise函数对象的 reject 方法
返回一个新的Promise对象 Promise.reject中不能再传入Promise
*/
    static reject = function (reason) {
      return new Promise((resolve, reject) => {
        reject(reason)
      })
    }

    /*
    Promise函数对象的 all 方法,接受一个promise类型的数组
    返回一个新的Promise对象
    所有promise成功则返回成功，只要有一个失败就失败
  */
    static all = function (promises) {
      // 成功的promise的值
      const values = new Array(promises.length)
      // 记录成功的promise的数目
      let count = 0
      return new Promise((resolve, reject) => {
        // 遍历promises，获取每个promise的结果
        promises.forEach((p, index) => {
          // 使用  Promise.resolve(p) 再包装一遍，以接收字面量参数
          Promise.resolve(p).then(
            value => {
              count++
              values[index] = value
              // 全部成功，返回所有成功的值的数组
              if (count === promises.length) {
                resolve(values)
              }

            },
            reason => {
              // 出现失败就返回失败
              resolve(reason)
            }
          )
        })

      })
    }

    /*
      Promise函数对象的 race 方法,接受一个promise类型的数组
      返回一个新的Promise对象
    */
    static race = function (promises) {
      return new Promise((resolve, reject) => {
        promises.forEach(p => {
          Promise.resolve(p).then(
            value => {
              resolve(value)
            },
            reason => {
              reject(reason)
            }
          )
        })
      })
    }

    /*
      resolveDelay：扩展工具方法，延迟返回结果
    */
    static resolveDelay = function (value, time) {
      return new Promise((resolve, reject) => {
        setTimeout(() => {
          if (value instanceof Promise) {
            value.then(resolve, reject)
          } else {
            resolve(value)
          }
        }, time)
      })
    }

    /*
    rejectDelay：扩展工具方法，延迟返回结果
  */
    static rejectDelay = function (reason, time) {
      return new Promise((resolve, reject) => {
        setTimeout(() => {
          reject(reason)
        }, time)
      })
    }

  }

  // 向外暴露 Promise 函数
  window.Promise = Promise
})(window)
```